api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,417 @@
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			heights = append(heights, epochLastBlock)
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			break
+		}
+	}
+}
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func (ss *EpochSync) getMaxPeerHeight(isBeacon bool) uint64 {
+	maxHeight := uint64(math.MaxUint64)
+	var (
+		wg   sync.WaitGroup
+		lock sync.Mutex
+	)
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			//debug
+			// utils.Logger().Debug().Bool("isBeacon", isBeacon).Str("peerIP", peerConfig.ip).Str("peerPort", peerConfig.port).Msg("[Sync]getMaxPeerHeight")
+			response, err := peerConfig.client.GetBlockChainHeight()
+			if err != nil {
+				utils.Logger().Warn().Err(err).Str("peerIP", peerConfig.ip).Str("peerPort", peerConfig.port).Msg("[Sync]GetBlockChainHeight failed")
+				ss.syncConfig.RemovePeer(peerConfig)
+				return
+			}
+			utils.Logger().Info().Str("peerIP", peerConfig.ip).Uint64("blockHeight", response.BlockHeight).
+				Msg("[SYNC] getMaxPeerHeight")
+
+			lock.Lock()
+			if response != nil {
+				if maxHeight == uint64(math.MaxUint64) || maxHeight < response.BlockHeight {
+					maxHeight = response.BlockHeight
+				}
+			}
+			lock.Unlock()
+		}()
+		return
+	})
+	wg.Wait()
+	return maxHeight
+}
+
+// ProcessStateSync processes state sync from the blocks received but not yet processed so far
+func (ss *EpochSync) ProcessStateSync(heights []uint64, bc *core.BlockChain, worker *worker.Worker) error {
+
+	var payload [][]byte
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		resp := peerConfig.GetClient().GetBlocksByHeights(heights, ss.selfip, ss.selfport)
+		if resp == nil {
+			return false
+		}
+		if len(resp.Payload) == 0 {
+			return false
+		}
+		payload = resp.Payload
+		return true
+	})
+
+	if len(payload) == 0 {
+		return errors.New("empty payload")
+	}
+
+	decoded := make([]*types.Block, 0, len(payload))
+	for idx, blockBytes := range payload {
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			return errors.Wrap(err, "failed decode")
+		}
+
+		if !block.IsLastBlockInEpoch() {
+			return errors.Errorf("block is not last index %d hash %s", idx, block.Hash().Hex())
+		}
+
+		decoded = append(decoded, block)
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
```

Comment:
I've added signature validation, but i'm not sure is it correct and enough? Maybe i should add some extra validation?

Reply:
Signature verification is enough and we can't validate other data since we don't have the consecutive blocks and state.